server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1356,19 +1366,14 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                             );
                         },
                         e -> {
-                            logger.warn("deleted snapshot failed - deleting files", e);
-                            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
-                                try {
-                                    deleteSnapshot(snapshot.getRepository(), snapshot.getSnapshotId().getName(), listener, true);
-                                } catch (SnapshotMissingException smex) {
-                                    logger.info(() -> new ParameterizedMessage(
-                                        "Tried deleting in-progress snapshot [{}], but it could not be found after failing to abort.",
-                                        smex.getSnapshotName()), e);
-                                    listener.onFailure(new SnapshotException(snapshot,
-                                        "Tried deleting in-progress snapshot [" + smex.getSnapshotName() + "], but it " +
-                                            "could not be found after failing to abort.", smex));
-                                }
-                            });
+                            if (abortedDuringInit) {
+                                logger.debug(() -> new ParameterizedMessage("Snapshot [{}] was aborted during INIT", snapshot), e);
+                                listener.onResponse(null);
+                            } else {
+                                logger.warn("deleted snapshot failed", e);
+                                listener.onFailure(
+                                    new SnapshotMissingException(snapshot.getRepository(), snapshot.getSnapshotId(), e));
```

Comment:
I'm not sure what cases this is supposed to cover. In particular, I'm wondering about the case where the current node failed (e.g. got disconnected from the rest of the cluster) and another master completed the snapshot. How are the listeners in `snapshotCompletionListeners` informed?

Reply:
> I'm not sure what cases this is supposed to cover.

The only way I see of getting here is the one in the linked test failure.
Master tried to finalize the snapshot and ran into an `IOException` (or other but I don't see which one).

That said ... you're right, on master fail-over we can leak the `snapshotCompletionListeners`  (urgh ... I wonder if that explain the odd test failure of a hanging snapshot once a month) I'll open another PR for that?

Reply:
https://github.com/elastic/elasticsearch/pull/54286 should do it here but I'd like a few hours of `SnapshotsResiliencyTests` to be sure :)

Reply:
By wrapping the original exception here, I wonder if we potentially turn a failing master (FailedToCommitClusterStateException / NotMasterException) into a `SnapshotMissingException`

Reply:
That's a good point ... I wonder if we should just pass those two exceptions (failed to commit/ not master) as they are. At this point, the delete has not in fact put anything into the cluster state aside from aborting the snapshot. So if we get here and run into on of those master fail-over exceptions, then retrying the delete request seems what we would actually want to happen right?

Reply:
Pushed https://github.com/elastic/elasticsearch/pull/54256/commits/2a2422b86923b496eb3d0e63bcb9c7aa2600d32c for the above